### ISSUE TITLE
Output from analysis tools should be a user friendly format

### ISSUE THREAD

**Issue by pwolfram** [MPAS-Dev/MPAS-Analysis#21]
The products of the analysis framework should be packaged in an easily usable (e.g. NetCDF / ASCII) / viewable format. (e.g. HTML file).

**Comment by pwolfram** [comment]
@xylar and @milenaveneziani, I would say that we are currently meeting this requirement, would you agree?  I'm wondering if we can close this issue for now.  Obviously, we need to ensure that this requirement is met moving forward too.

**Comment by xylar** [comment]
@pwolfram, we are currently outputting images without any further documentation, formatting or metadata so, no, I don't think we're complying with the intent of this requirement yet.

**Comment by pwolfram** [comment]
@xylar, I definitely agree we could be doing better.  However, in terms of ACME we have the output webpage that provides some minimal user-interfaces and explanation.  It could be viewed as a minimalist meeting of this requirement although it is probably a stretch.

I suppose the question then is how should we refine this requirement so that we can make this actionable?  This is obviously not a week-timescale critical issue but is something we should address sooner than later.

**Comment by xylar** [comment]
@pwolfram 

>  However, in terms of ACME we have the output webpage that provides some minimal user-interfaces and explanation.

I don't think that's relevant to this repository and its requirements, since this is not simply meant to be a helper for ACME analysis.  If that functionality were moved (or copied) here, we might be able to say that we were meeting the requirement, though.

>  I suppose the question then is how should we refine this requirement so that we can make this actionable? 

I guess this is my issue in general with putting requirements for the design document as "issues" in the repo.  Typically they are much broader than a single issue.

This should be broken into a series of steps.  What those steps are should be discussed, but something like
* writing out processed data in NetCDF files (with appropriate metadata for both provenance and documentation)
* producing HTML to more effectively display the output images (including provenance and other relevant metadata)
* ...

**Comment by pwolfram** [comment]
Thanks for clarifying @xylar, you are right that in terms of this repository we are not meeting this requirement.  I agree we need a punch-list of steps that are designed to fully address this requirement.

**Comment by milenaveneziani** [comment]
I would say let us start with breaking down this issue in the two points that @xylar made above.
I would prioritize the html interface, since that will make our life much easier in terms of debugging MPAS-Analysis scripts (prior to moving things to the ACME repo). We have discussed this at the ACME meeting with the workflow team: @pwolfram, do you have a sense of where that discussion left us and whether we should pursue that road now?

**Comment by pwolfram** [comment]
If I recall correctly @milenaveneziani, we needed more info from the workflow team to know how to use their existing self-documenting workflow.  Essentially, this boils down to the correct format to use for meta-data that can be parsed and turned into some type of workflow viewer, if I understand correctly.  This is something we should pursue for the long term.  The question is of course priority and that should be assessed in our road map.

**Comment by milenaveneziani** [comment]
ok, let us talk about this next week.
Personally, I think it would be nice to bump up the priority of this, if it does not translate into a substantial time sink.

**Comment by xylar** [comment]
cc @milenaveneziani @mark-petersen @stephenprice, @matthewhoffman @pwolfram @vanroekel @JeremyFyke @darincomeau

Here is a mockup website from MPAS-Analysis.
http://portal.nersc.gov/project/acme/xylar/mpas_analysis_website_mockup/
I would appreciate feedback before I start automating the process of generating sites like this.

**Comment by xylar** [comment]
In particular:
* Do you prefer having 6 separate webpages like I have here or would you prefer to have everything on one page?
* To what extent is is helpful to have separate "galleries" and to what extent would you just like to page through all the images with the left and right arrow keys?

**Comment by xylar** [comment]
One option would be to have separate webpages for each component but not for individual types of analysis (so currently 2 pages, one for ocean and one for sea ice).

**Comment by mark-petersen** [comment]
@xylar, this is fantastic, thanks!  I think the top page is perfect, and having a single top page as a table of contents is the right amount.  

On the sub pages, I would prefer to have the full image shown rather than a tight square.  The zoomed in image has the effect, here:
http://portal.nersc.gov/project/acme/xylar/mpas_analysis_website_mockup/seaIce/climatologyMaps/index.html
of showing only the observations because it is the middle plot, so that is really not useful to assess the model state at a glance. I know the squares line up for a more organized look, but the result is less functional.

**Comment by xylar** [comment]
@mark-petersen, the thumbnail generation was automatic and could easily be made to focus on the top of the image rather than the middle.  

My concern about showing the full image is both that the aspect ratio of some images is *quite* awkward for a gallery and that the images have various aspect ratios, leading to serious formatting issues when displaying them as a gallery.  But I'm definitely willing to try it out and see how it goes.

**Comment by xylar** [comment]
Here's an alternative version where there is one page per component:
http://portal.nersc.gov/project/acme/xylar/mpas_analysis_website_mockup2/

**Comment by xylar** [comment]
@mark-petersen, I tried doing thumbnails as you suggested (with the whole image).  It works better than I had feared.  Take a look at the second version of the website in the link above at tell me what you think.  You may need to do shift-F5 (or the mac equivalent) to force the webpage to fully refresh if you've already looked at the site before my recent changes.

**Comment by kevans32** [comment]
Hey guys- before you spend tons of time on making a user friendly website, we are going to be hooking this up to existing web interfaces we have with LIVV and A-PRIME at some point, and could probably accelerate that task if you like. Would that be useful?

**Comment by kevans32** [comment]
That said, the APRIME is fairly minimal itself, so the only benefit to moving this stuff there is to connect it to the atmosphere and a system the coupled team is already running.

**Comment by xylar** [comment]
@kevans32, we definitely need some HTML that's appropriate for running standalone MPAS-Analysis. 
The current A-PRIME html generation doesn't really meet our needs, in that you have to click on a bunch of links to see the output and it doesn't include new analysis capabilities that we're adding (very quickly) on this side.

If the work we do here can dovetail with LIVV and A-PRIME needs, that would be great!  We're keen collaborate!  But we need this up and running for us very soon, preferably by early next week, since we are doing a lot of runs and struggling to view the output from them (currently just a directory of images).

How would you feel if we proceed here and try to merge our efforts in the not-to-distant future?

**Comment by xylar** [comment]
Also, your feedback and also from others on the coupled team on the web design would be most appreciated, especially since that would help later on when we combine efforts.

**Comment by kevans32** [comment]
Agreed A-PRIME does not suit, and I do not know if Dave wants us to extend it to to be more useful for all components long term. We are tasked to connect with MPAS-Analysis as part of the new SciDAC to extend the ice sheet validation. I see that you guys are doing tons of work there on and will not get in your way. But I think we have some tools you may like. How about we talk at SIAM about what is useful to you over the next several months and task it out?

**Comment by xylar** [comment]
Sounds great!

**Comment by xylar** [comment]
@akturner, forgot to cc you but would appreciate your feedback.

**Comment by milenaveneziani** [comment]
@xylar: this is great!
My preference is for the second mockup, with components separated and the full figure thumbnails. I think that the visualization style is very useful. You can imagine that in the future we'll have a similar html page for analysis that compares two different runs together, instead of one run wrt observations, and this format would work very well for that kind of visualization as well.
A few minor things that I think would improve it:
1) an index of sections at the top for ease of browsing (climatologies, time series, mht/moc, nino3.4 diags, etc)
2) in some cases the figure titles are big enough to be visible in the thumbnails, but in other cases they are not. When comparing different things in the thumbnail, it is nice to see a title for each column of figures. We could either change the original png's or add a title in the thumbnails.
3) I would either not show the Bootstrap-derived climatologies for sea-ice, or add a clear title so that the differences between the different figures are clear. Also, we do not need to show the MHT vs depth figure.

again, thanks for working on this.

**Comment by xylar** [comment]
@milenaveneziani, these are great suggestions.  I'm working on adding them to mockup2.

**Comment by xylar** [comment]
> an index of sections at the top for ease of browsing

I'm trying to do this with images because I prefer visual (rather than text-based) navigating

> We could either change the original png's or add a title in the thumbnails.

I think changing the original pngs so the thumbnails have readable titles is an awkward but possible solution.  The titles would need to be simplified so they could be made bigger.  In general, I'm in favor of this.

It won't be easy to add titles to the thumbnails, especially in a way that makes them readable.  

To be honest, I'm not sure how important this is.  I presume one can click on an arbitrary image in the gallery and see what it is.  It's quite easy to browse through the gallery to find what you were after.

> I would either not show the Bootstrap-derived climatologies for sea-ice, or add a clear title so that the differences between the different figures are clear. 

I think this is more of an issue with the analysis task that produced these images than with the webpage.  Perhaps we should make and issue for this?  One option here is that I could do separate galleries for Bootstrap and NASA Team results.  I am not in favor of excluding from the webpage any images we produce.  It makes things more complicated because I have to somehow know which images *don't* get included when the webpage gets generated.  I also think it would be confusing for people digging into the analysis that we produce image but don't actually use them on the webpages.  

Perhaps we need to add and issue for removing Bootstrap output or turn it off by default?

> Also, we do not need to show the MHT vs depth figure.

Again, I think if we produce the image, we should include it on the webpage.  If you think we shouldn't produce the image, it would be best to create and issue where we remove it or make it "off" by default.

**Comment by milenaveneziani** [comment]
> It won't be easy to add titles to the thumbnails, especially in a way that makes them readable.
>
>To be honest, I'm not sure how important this is. I presume one can click on an arbitrary image in the gallery and see what it is. It's quite easy to browse through the gallery to find what you were after.

I saw the thumbnails useful for comparing all plots side by side, without having to click on the single image to see the details (which of corse one does if you do want to see the details). So having big titles in that situation helps a lot.

About the Bootstrap/MHT plots: I would prefer to keep producing Bootstrap images (as also suggested by @akturner a long time ago), but grouping them together (say NasaTeam JFM/JAS, Bootstrap JFM/JAS, similar to how we have them in A-prime). While I would remove the MHT with depth from the analysis, unless other people think they are useful.

**Comment by xylar** [comment]
> I saw the thumbnails useful for comparing all plots side by side, without having to click on the single image to see the details

Another option is to have a caption for each image, perhaps in a rather small font.  This makes the view less compact but not by much if the caption just has a season, a depth or something like that.

**Comment by xylar** [comment]
> I would prefer to keep producing Bootstrap images ... but grouping them together (say NasaTeam JFM/JAS, Bootstrap JFM/JAS, similar to how we have them in A-prime).

I have done this in the latest iteration:
http://portal.nersc.gov/project/acme/xylar/mpas_analysis_website_mockup2/seaIce/index.html

Please let me know what you think.

**Comment by xylar** [comment]
@milenaveneziani, captions do seem to work rather well.  See what you think of the captions I have so far on the ocean site:
http://portal.nersc.gov/project/acme/xylar/mpas_analysis_website_mockup2/ocean/index.html

I haven't completed this work because it's rather tedious but I think you get the idea of how it would look.

**Comment by milenaveneziani** [comment]
It all looks very nice to me, @xylar!

**Comment by toddringler** [comment]
This is beautiful.

**Comment by stephenprice** [comment]
This is fantastic Xylar. I'm sure there's some format tweaking we can do down the line, but it looks great so far. This will be super handy.

**Comment by xylar** [comment]
> I'm sure there's some format tweaking we can do down the line.

Yep, for sure.  The main things that will be a little harder to tweak later would be things like how many pages are generated (e.g. one per component vs. one per analysis type) so that's where the feedback so far has been especially helpful.

**Comment by stephenprice** [comment]
Nominally, one page per component would seem to make sense to me. I can see some grey areas that popping up eventually (Does ice shelf melting below to land ice or ocean? Do icebergs belong to land ice, sea ice, or ocean?). But addressing those as they come up makes more sense to me than grouping everything by say, time series vs. spatial maps.

**Comment by xylar** [comment]
I think this could be handled easily by having a page for land ice/ocean interactions, too. For niw, they seem to belong to the ocean, though.

**Comment by xylar** [comment]
@stephenprice, thinking some more about this issue at more the task level rather than the html level, MPAS-Analysis currently associates each task with a single MPAS component whose output files it expects to get its data from.  Do we foresee individual analysis tasks that uses history files from multiple components (e.g. comparing melt rates as seen by the ocean to those seen by the land ice, or budgets at the interface between components)?  I could imagine a few like these, and I guess we would need to address that case when we first encounter it.

For your example of ice shelf melt, I would think it could belong to either land ice or ocean, depending on where the history files come from, but that this would be the determining factor.  For the case of icebergs, I would think they would likely belong to the sea ice component (or just be a component of their own), given that the history files are likely to be written out by the sea-ice model.

**Comment by stephenprice** [comment]
@xylar - For now I don't think we should make this more complicated than it needs to be, so I'd be fine with just assigning things to a particular component. For now, it certainly makes sense to assign melt fluxes from the ice shelf to the ocean component (overall, this is a freshwater source term for the ocean). There will eventually be plenty of other ice-sheet specific analyses that can be assigned to the ice sheet component.
